File: native/ios/VueNativeCore/Sources/VueNativeCore/Modules/BackgroundTaskModule.swift
#if canImport(UIKit)
import UIKit
import BackgroundTasks

/// Native module for scheduling background tasks using BGTaskScheduler.
///
/// Methods:
///   - scheduleTask(taskId, type, options) — schedule a background task
///   - cancelTask(taskId) — cancel a specific task
///   - cancelAllTasks() — cancel all scheduled tasks
///   - completeTask(taskId) — signal task completion from JS
///
/// Events:
///   - background:taskExecute — fired when a background task runs, payload: { taskId }
@available(iOS 13.0, *)
final class BackgroundTaskModule: NativeModule {
    var moduleName: String { "BackgroundTask" }
    private weak var bridge: NativeBridge?

    /// Track active tasks so we can call setTaskCompleted from JS
    private var activeTasks: [String: BGTask] = [:]

    /// Registered task identifiers
    private var registeredIdentifiers: Set<String> = []

    init(bridge: NativeBridge) {
        self.bridge = bridge
    }

    func invoke(method: String, args: [Any], callback: @escaping (Any?, String?) -> Void) {
        switch method {
        case "scheduleTask":
            guard args.count >= 2,
                  let taskId = args[0] as? String,
                  let type = args[1] as? String else {
                callback(nil, "scheduleTask: missing taskId or type")
                return
            }
            let options = args.count >= 3 ? args[2] as? [String: Any] : nil
            scheduleTask(taskId: taskId, type: type, options: options ?? [:], callback: callback)

        case "cancelTask":
            guard let taskId = args.first as? String else {
                callback(nil, "cancelTask: missing taskId")
                return
            }
            BGTaskScheduler.shared.cancel(taskIdentifier: taskId)
            callback(nil, nil)

        case "cancelAllTasks":
            BGTaskScheduler.shared.cancelAllTaskRequests()
            callback(nil, nil)

        case "completeTask":
            guard let taskId = args.first as? String else {
                callback(nil, "completeTask: missing taskId")
                return
            }
            let success = (args.count >= 2 ? args[1] as? Bool : nil) ?? true
            if let task = activeTasks[taskId] {
                task.setTaskCompleted(success: success)
                activeTasks.removeValue(forKey: taskId)
            }
            callback(nil, nil)

        case "registerTask":
            guard let taskId = args.first as? String else {
                callback(nil, "registerTask: missing taskId")
                return
            }
            registerTaskHandler(taskId: taskId)
            callback(nil, nil)

        default:
            callback(nil, "BackgroundTaskModule: Unknown method '\(method)'")
        }
    }

    // MARK: - Private

    private func registerTaskHandler(taskId: String) {
        guard !registeredIdentifiers.contains(taskId) else { return }
        registeredIdentifiers.insert(taskId)

        BGTaskScheduler.shared.register(forTaskWithIdentifier: taskId, using: nil) { [weak self] task in
            self?.handleTaskExecution(task: task)
        }
    }

    private func handleTaskExecution(task: BGTask) {
        let taskId = task.identifier
        activeTasks[taskId] = task

        task.expirationHandler = { [weak self] in
            self?.activeTasks[taskId]?.setTaskCompleted(success: false)
            self?.activeTasks.removeValue(forKey: taskId)
        }

        // Notify JS that the task is executing
        let bridge = bridge
        DispatchQueue.main.async {
            bridge?.dispatchGlobalEvent("background:taskExecute", payload: ["taskId": taskId])
        }
    }

    private func scheduleTask(taskId: String, type: String, options: [String: Any], callback: @escaping (Any?, String?) -> Void) {
        // Ensure task handler is registered
        registerTaskHandler(taskId: taskId)

        let request: BGTaskRequest
        if type == "processing" {
            let processingRequest = BGProcessingTaskRequest(identifier: taskId)
            processingRequest.requiresNetworkConnectivity = options["requiresNetworkConnectivity"] as? Bool ?? false
            processingRequest.requiresExternalPower = options["requiresExternalPower"] as? Bool ?? false
            request = processingRequest
        } else {
            // Default: app refresh
            request = BGAppRefreshTaskRequest(identifier: taskId)
        }

        if let earliestBeginDate = options["earliestBeginDate"] as? Double {
            request.earliestBeginDate = Date(timeIntervalSince1970: earliestBeginDate / 1000.0)
        }

        do {
            try BGTaskScheduler.shared.submit(request)
            callback(nil, nil)
        } catch {
            callback(nil, "Failed to schedule task: \(error.localizedDescription)")
        }
    }

    func invokeSync(method: String, args: [Any]) -> Any? { nil }
}
#endif

File: native/android/VueNativeCore/src/main/kotlin/com/vuenative/core/Modules/BackgroundTaskModule.kt
package com.vuenative.core

import android.content.Context
import android.util.Log
import androidx.work.*
import java.util.concurrent.TimeUnit

/**
 * Native module for scheduling background tasks using WorkManager.
 *
 * Methods:
 *   - scheduleTask(taskId, type, options) — schedule a one-time or periodic task
 *   - cancelTask(taskId) — cancel a specific task
 *   - cancelAllTasks() — cancel all scheduled tasks
 *   - completeTask(taskId) — no-op on Android (Worker handles completion)
 *   - registerTask(taskId) — no-op on Android (registration is automatic)
 *
 * Events:
 *   - background:taskExecute — fired when a background task runs, payload: { taskId }
 */
class BackgroundTaskModule : NativeModule {
    override val moduleName = "BackgroundTask"
    private var appContext: Context? = null
    private var bridgeRef: NativeBridge? = null

    override fun initialize(context: Context, bridge: NativeBridge) {
        appContext = context.applicationContext
        bridgeRef = bridge
        // Store bridge reference for the worker
        VueNativeWorker.bridgeRef = bridge
    }

    override fun invoke(method: String, args: List<Any?>, bridge: NativeBridge, callback: (Any?, String?) -> Unit) {
        val ctx = appContext ?: run {
            callback(null, "BackgroundTask not initialized")
            return
        }

        when (method) {
            "scheduleTask" -> {
                val taskId = args.getOrNull(0)?.toString()
                    ?: run {
                        callback(null, "scheduleTask: missing taskId")
                        return
                    }
                val type = args.getOrNull(1)?.toString() ?: "refresh"
                val options = args.getOrNull(2) as? Map<*, *> ?: emptyMap<String, Any>()
                scheduleTask(ctx, taskId, type, options, callback)
            }
            "cancelTask" -> {
                val taskId = args.getOrNull(0)?.toString()
                    ?: run {
                        callback(null, "cancelTask: missing taskId")
                        return
                    }
                WorkManager.getInstance(ctx).cancelUniqueWork(taskId)
                callback(null, null)
            }
            "cancelAllTasks" -> {
                WorkManager.getInstance(ctx).cancelAllWork()
                callback(null, null)
            }
            "completeTask" -> {
                // On Android, WorkManager Workers handle their own completion
                callback(null, null)
            }
            "registerTask" -> {
                // No-op on Android — registration happens at schedule time
                callback(null, null)
            }
            else -> callback(null, "BackgroundTaskModule: Unknown method '$method'")
        }
    }

    private fun scheduleTask(
        ctx: Context,
        taskId: String,
        type: String,
        options: Map<*, *>,
        callback: (Any?, String?) -> Unit
    ) {
        try {
            val constraintsBuilder = Constraints.Builder()

            val requiresNetwork = options["requiresNetworkConnectivity"] as? Boolean ?: false
            if (requiresNetwork) {
                constraintsBuilder.setRequiredNetworkType(NetworkType.CONNECTED)
            }

            val requiresCharging = options["requiresExternalPower"] as? Boolean ?: false
            if (requiresCharging) {
                constraintsBuilder.setRequiresCharging(true)
            }

            val constraints = constraintsBuilder.build()

            val inputData = Data.Builder()
                .putString("taskId", taskId)
                .build()

            if (type == "processing") {
                // Periodic work for long-running/processing tasks
                val intervalMinutes = (options["interval"] as? Number)?.toLong() ?: 15L
                val request = PeriodicWorkRequestBuilder<VueNativeWorker>(
                    intervalMinutes, TimeUnit.MINUTES
                )
                    .setConstraints(constraints)
                    .setInputData(inputData)
                    .addTag("vue-native-bg-$taskId")
                    .build()

                WorkManager.getInstance(ctx).enqueueUniquePeriodicWork(
                    taskId,
                    ExistingPeriodicWorkPolicy.REPLACE,
                    request
                )
            } else {
                // One-time work for refresh tasks
                val requestBuilder = OneTimeWorkRequestBuilder<VueNativeWorker>()
                    .setConstraints(constraints)
                    .setInputData(inputData)
                    .addTag("vue-native-bg-$taskId")

                val delayMs = (options["earliestBeginDate"] as? Number)?.toLong()
                if (delayMs != null) {
                    val now = System.currentTimeMillis()
                    val delay = delayMs - now
                    if (delay > 0) {
                        requestBuilder.setInitialDelay(delay, TimeUnit.MILLISECONDS)
                    }
                }

                WorkManager.getInstance(ctx).enqueueUniqueWork(
                    taskId,
                    ExistingWorkPolicy.REPLACE,
                    requestBuilder.build()
                )
            }

            callback(null, null)
        } catch (e: Exception) {
            callback(null, "Failed to schedule task: ${e.message}")
        }
    }

    override fun destroy() {
        VueNativeWorker.bridgeRef = null
    }
}

/**
 * Worker that fires background:taskExecute events back to the JS bridge.
 */
class VueNativeWorker(context: Context, params: WorkerParameters) : Worker(context, params) {
    companion object {
        @Volatile
        var bridgeRef: NativeBridge? = null
    }

    override fun doWork(): Result {
        val taskId = inputData.getString("taskId") ?: return Result.failure()

        try {
            bridgeRef?.dispatchGlobalEvent(
                "background:taskExecute",
                mapOf("taskId" to taskId)
            )
        } catch (e: Exception) {
            Log.e("VueNative", "BackgroundTask execution failed", e)
            return Result.failure()
        }

        return Result.success()
    }
}
